clippy_lints/src/loops.rs

```diff
@@ -590,9 +587,16 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for Loops {
             }
         }
 
-        // check for while loops which conditions never change
-        if let ExprKind::While(ref cond, _, _) = expr.node {
-            check_infinite_loop(cx, cond, expr);
+        if_chain! {
+            if let ExprKind::Loop(block, _, LoopSource::While) = &expr.node;
+            if let Block { expr: Some(expr), .. } = &**block;
+            if let ExprKind::Match(cond, arms, MatchSource::WhileDesugar) = &expr.node;
+            if let ExprKind::DropTemps(cond) = &cond.node;
```

Comment:
This should be a `while_desugar` in `utils` like the `if_let_desugar` and `for_desugar` we already have